future-apps/staking/contracts/Staking.sol

```diff
@@ -0,0 +1,221 @@
+pragma solidity 0.4.18;
+
+import "./ERCStaking.sol";
+
+import "@aragon/os/contracts/apps/AragonApp.sol";
+import "@aragon/os/contracts/lib/zeppelin/token/ERC20.sol";
+
+
+contract Staking is ERCStaking, AragonApp {
+  struct Account {
+    uint256 amount;
+    Lock[] locks;
+  }
+
+  struct Lock {
+    uint256 amount;
+    Timespan timespan;
+    address unlocker;
+    bytes32 metadata;
+  }
+
+  struct Timespan {
+    //uint64 start; // exclusive
+    uint64 end;     // inclusive
+    TimeUnit unit;
+  }
+
+  enum TimeUnit { Blocks, Seconds }
+
+  ERC20 public stakingToken;
+  bytes public stakeScript;
+  bytes public unstakeScript;
+  bytes public lockScript;
+
+  mapping (address => Account) accounts;
+
+  // TODO: figure out a way to get lock from metadata, given changing lock ids
+  // mapping (bytes32 => LockLocation) lockByMetadata;
+
+  event Locked(address indexed account, uint256 lockId, bytes32 metadata);
+  event Unlocked(address indexed account, address indexed unlocker, uint256 oldLockId);
+
+  event MovedTokens(address indexed from, address indexed to, uint256 amount);
+
+  bytes32 constant STAKE_ROLE = keccak256("STAKE_ROLE");
+  bytes32 constant UNSTAKE_ROLE = keccak256("UNSTAKE_ROLE");
+  bytes32 constant LOCK_ROLE = keccak256("LOCK_ROLE");
+  bytes32 constant GOD_ROLE = keccak256("GOD_ROLE");
+
+  modifier checkUnlocked(uint256 amount) {
+    require(unlockedBalanceOf(msg.sender) >= amount);
+    _;
+  }
+
+  // TODO: Implement forwarder interface
+
+  function initialize(ERC20 _stakingToken, bytes _stakeScript, bytes _unstakeScript, bytes _lockScript) onlyInit external {
+    stakingToken = _stakingToken;
+    stakeScript = _stakeScript;
+    unstakeScript = _unstakeScript;
+    lockScript = _lockScript;
+
+    initialized();
+  }
+
+  function stake(uint256 amount, bytes data) authP(STAKE_ROLE, arr(amount)) public {
+    stakeFor(msg.sender, amount, data);
+  }
+
+  function stakeFor(address acct, uint256 amount, bytes data) authP(STAKE_ROLE, arr(amount)) public {
+    // From needs to be msg.sender to avoid token stealing by front-running
+    require(stakingToken.transferFrom(msg.sender, this, amount));
+    processStake(acct, amount, data);
+
+    if (stakeScript.length > 0) {
+      runScript(stakeScript, data, new address[](0));
+    }
+  }
+
+  function processStake(address acct, uint256 amount, bytes data) internal {
+    accounts[acct].amount += amount; // overflow check in token contract
+    assert(accounts[acct].amount >= amount);
```

Comment:
What do you mean with the comment one line above? Aren't you checking overflow here?
Shouldn't we use SafeMath lib like the rest of the apps?
Do we allow amount = 0?